Snapshot number and utxo hash need not be in the redeemer on close/contest

ffakenz opened the pull request:
⛰️ Remove snapshot number and utxo hash from Contest and Close redeemers

* the output datum needs to include them and we can check using that
* only the signature is enough

⛰️ BTW this improves the coverage for the case where we close with the initialSnapshot

To check before merging:
* [x] CHANGELOG is up to date
* [x] Documentation is up to date

ffakenz committed:
Move snapshot number from Contest redeemer to Closed datum in Head script

ffakenz committed:
Move snapshot number from Close redeemer to Closed datum in Head script

ffakenz committed:
Move utxo hash from Contest redeemer to Closed datum in Head script

ffakenz committed:
Move utxo hash from Close redeemer to Closed datum in Head script

ffakenz committed:
remove utxo hash attribute from close redeemer

ffakenz committed:
minor refactor to upgrade continuingDatum from local to a helper function

ffakenz committed:
minor

ffakenz committed:
HACK: reduce the max supported number of parties to reduce the size of fanout

ffakenz committed:
Update changelog

v0d1ch commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -258,8 +256,13 @@ checkClose ctx parties initialUtxoHash snapshotNumber closedUtxoHash sig cperiod
   checkSnapshot
-    | snapshotNumber == 0 =
+    | closedSnapshotNumber == 0 =
       let expectedOutputDatum =
```
I remember vaguely @ch1bo mentioned something about not needing to construct the expected datum here but I could be crazy as well.

ch1bo replied:
Yeah.. this code is decoding the `continuingDatum` twice.

Once to get `closedSnapshotNumber` and `closedUtxoHash`, then it builds the expected datum value and compares it against the `continuingDatum`.

Simpler would be to just do the checks we care about. That is
* if `closedSnapshotNumber >= 0`, `verifySnapshotSignature parties closedSnapshotNumber closedUtxoHash sig`
* else (== 0 or negative), `closedUtxoHash == initialUtxoHash`

That should also make the code smaller and more efficient.

v0d1ch commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -477,6 +482,19 @@ mustContinueHeadWith ScriptContext{scriptContextTxInfo = txInfo} headAddress cha
 
+continuingDatum :: ScriptContext -> Datum
+continuingDatum ctx =
+  case ownDatum of
```
There is a similar function here https://github.com/input-output-hk/hydra/blob/master/hydra-plutus/src/Hydra/Contract/Head.hs#L480 
Perhaps we can extract it to `Util` module and use at both places?

ch1bo replied:
I would rather use `findTxOutDatum` to define this whole thing:
```
continuingDatum ctx =
  case getContinuingOutputs ctx of
      [o] -> findTxOutDatum o
      _ -> traceError "expected only one continuing output"
```

v0d1ch reviewed (approved):
Looks good to me overall @ffakenz, I just have a concern @ch1bo will be able to resolve which is related to constructing the expected datum.

ch1bo commented on CHANGELOG.md:
```diff
@@ -60,6 +60,8 @@ changes.
   `NotEnoughFuel` when utxo was not found. Now `NotEnoughFuel` is used when
   there is not enough fuel and `NoFuelUTXOFound` when utxo was not to be found.
 
+- Remove snapshot number and utxo hash from Close and Contest reedemer as they were already present in Closed datum.
```
Could: move this already under the other `**BREAKING** hydra-plutus ` changes.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -160,20 +174,36 @@ genCloseMutation (tx, _utxo) =
-        mutatedSnapshotNumber <- arbitrarySizedNatural `suchThat` (\n -> n /= healthySnapshotNumber && n > 0)
-        pure (closeRedeemer mutatedSnapshotNumber $ healthySignature healthySnapshotNumber)
-    , SomeMutation MutateSnapshotToIllFormedValue . ChangeHeadRedeemer <$> do
+        closeRedeemer <$> (arbitrary :: Gen (MultiSignature (Snapshot Tx)))
```
Can we address the `FIXME` above now - is it still relevant? It seems like the redeemer became trivial now and we should not use this function per the comment.

ch1bo replied:
I think yes, just inline it.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -160,20 +174,36 @@ genCloseMutation (tx, _utxo) =
+            , parties = healthyOnChainParties
+            , contestationDeadline = posixFromUTCTime healthyContestationDeadline
+            , headId = toPlutusCurrencySymbol Fixture.testPolicyId
+            }
```
Should: use the `healthyCloseDatum` and only change `snapshotNumber = mutatedSnapshotNumber`

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -160,20 +174,36 @@ genCloseMutation (tx, _utxo) =
-    , SomeMutation MutateSnapshotToIllFormedValue . ChangeHeadRedeemer <$> do
+        closeRedeemer <$> (arbitrary :: Gen (MultiSignature (Snapshot Tx)))
+    , SomeMutation MutateSnapshotNumberButNotSignature . ChangeHeadDatum <$> do
+        (UnsafeSnapshotNumber mutatedSnapshotNumber) <- arbitrarySizedNatural `suchThat` (\n -> n /= healthySnapshotNumber && n > 0)
```
Must: not unwrap using `UnsafeSnapshotNumber`, but instead use one of the "not unsafe" instances.. in fact, you are already using `Integral` with `arbitrarySizedNatural` and `toInteger` below. So change it to:
```suggestion
        mutatedSnapshotNumber <- arbitrarySizedNatural `suchThat` (\n -> n /= healthySnapshotNumber && n > 0)
```

ch1bo reviewed (changes requested):
As @v0d1ch mentioned, the on-chain code can and should be simplified before we merge this.

v0d1ch committed:
Simplify on-chain code a bit

v0d1ch committed:
Alter continuingDatum function

v0d1ch committed:
Use healthyCloseDatum to construct Close datum

v0d1ch committed:
Inline closeRedeemer

pgrange commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -478,6 +464,13 @@ mustContinueHeadWith ScriptContext{scriptContextTxInfo = txInfo} headAddress cha
+    [o] -> findTxOutDatum scriptContextTxInfo o
+    _ -> traceError "expected only one continuing output"
+{-# INLINEABLE continuingDatum #-}
+
```
```suggestion
closeContextOutDatum :: ScriptContext -> (SnapshotNumber, Hydra.Contract.HeadState.Hash)
closeContextOutDatum ctx =
  case fromBuiltinData @DatumType $ getDatum (continuingDatum ctx) of
    Just Closed{snapshotNumber, utxoHash} -> (snapshotNumber, utxoHash)
    _ -> traceError "wrong state in output datum"
{-# INLINEABLE closeContextOutDatum #-}
```

pgrange commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -332,6 +312,12 @@ checkContest ctx@ScriptContext{scriptContextTxInfo} contestationDeadline parties
+    case fromBuiltinData @DatumType $ getDatum (continuingDatum ctx) of
+      Just Closed{snapshotNumber, utxoHash} -> (snapshotNumber, utxoHash)
+      _ -> traceError "wrong state in output datum"
+
```
```suggestion

  (contestSnapshotNumber, contestUtxoHash) = closeContextOutDatum ctx
```

pgrange commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -258,28 +256,14 @@ checkClose ctx parties initialUtxoHash snapshotNumber closedUtxoHash sig cperiod
+    case fromBuiltinData @DatumType $ getDatum (continuingDatum ctx) of
+      Just Closed{snapshotNumber, utxoHash} -> (snapshotNumber, utxoHash)
+      _ -> traceError "wrong state in output datum"
+
```
```suggestion
  (closedSnapshotNumber, closedUtxoHash) = closeContextOutDatum ctx
  
```

pgrange reviewed (changes requested):
Code duplication when fetching datum from the tx in close and contest. See proposed code change to remove duplication.

Side note: I'm having trouble with naming this function. Maybe you can find a better or, maybe, this is a smell that something's odd here and could be improved but I can't find it.

pgrange commented:
Some tests are failing since what looks like the first commits of the branch.

~~This [build](https://github.com/input-output-hk/hydra/actions/runs/3933023667/jobs/6726273632) fails since this [commit](https://github.com/input-output-hk/hydra/commit/4fc64f7605387b1cc3560a71ace09f1e28172bab) (which has been overwritten since then by some rebase I guess).~~

This [build](https://github.com/input-output-hk/hydra/actions/runs/3966763358/jobs/6797954550) make me believe this [commit](https://github.com/input-output-hk/hydra/commit/a3174f086c1974768eafa5624b3f3d8ecfad8717) introduced the issue.

I'm moving this P.R. back to in progress to investigate and fix.

pgrange committed:
Revert Simplify on-chain code a bit

abailly committed:
Check close contestation deadline explicitly

Also refine mutation to ensure it trips the right validation

ch1bo committed:
Rename ChangeHeadDatum -> ChangeInputHeadDatum

We also found a mutation test that fails for the wrong reason and hence
renamed this basic mutation operator to be hopefully clearer.

ch1bo committed:
Add a new mutation test scenario for closing with the initial snapshot

This is another interesting case which want to explore in the mutation
testing suite and it's easier to start with a new healthy tx of this
case.

Later (ideally) we would have generators for each case and have our
mutations defined generically on the generated cases.

pgrange committed:
FIX validator must check contestation period

When closing with the initial utxo we must check the contestation period

pgrange committed:
Optimize plutus code

Reducing to only two cases should reduce on-chain code size

ch1bo committed:
Benchmark fanout with 20+ assets

pgrange committed:
Reducing transaction size

That's not super clever to do it that way but we now that, at some
point, we're going to spend some time to really improve that with
short and precise error codes or something.

pgrange committed:
This does not change test output so keeping it that way

pgrange commented on hydra-node/hydra-node.cabal:
```diff
@@ -363,5 +364,5 @@ test-suite tests
     , yaml
 
-  build-tool-depends: hspec-discover:hspec-discover
+  build-tool-depends: hspec-discover:hspec-discover -any
```
I can't find any documentation of this `-any` option. What is it for?

ch1bo replied:
It was added by my formatting tool and looks like a version indicator?

pgrange replied:
I don't know, I did not find any trace of this _-any_ flag in spec-discover doc nor code 😢

pgrange committed:
Improve mutation readability

Use builder pattern to create mutation.

ch1bo commented on hydra-node/exe/tx-cost/TxCost.hs:
```diff
@@ -190,8 +190,8 @@ computeAbortCost =
-  interesting <- catMaybes <$> mapM compute [1, 2, 3, 5, 10, 20]
-  limit <- maybeToList . getFirst <$> foldMapM (fmap First . compute) [100, 99 .. 21]
+  interesting <- catMaybes <$> mapM compute [1, 2, 3, 5, 10, 50]
+  limit <- maybeToList . getFirst <$> foldMapM (fmap First . compute) [100, 99 .. 51]
```
@pgrange Why did you revert this? This DID have an effect as it was running the transaction cost benchmarks with <50 assets to see what the upper bound right now is (roughly 49).

pgrange replied:
That was my next item in my todo-list: ask you what's the purpose of this change.

As of yesterday, I've noticed that the test failed for size problems and then you pushed this change so I guessed that it was related to that and I wanted to check that we could pass the test without it.

Now I'm happy to understand what's the purpose of the change?

ch1bo replied:
It will update the benchmark on this PR to get a ballpark estimate of what the maximum number of assets supported to fan out is right now. For example:


| UTxO  | Tx size | % max Mem | % max CPU | Min fee ₳ |
| :---- | ------: | --------: | --------: | --------: |
...
| 49| 16374 | 83.44 | 46.74 | 1.88 |

So this can be used to have the failing test (in StateSpec) not try to fit too many assets into one 16kB transaction - there is a hard-coded upper bound used to generate tests https://github.com/input-output-hk/hydra/blob/08e155bdaf871a72bf3aedf6ce73796a5f6b9298/hydra-node/test/Hydra/Chain/Direct/StateSpec.hs#L392.

pgrange replied:
> So this can be used to have the failing test (in StateSpec)

OK then maybe that's the point. I did not observer any failing test in the last push. But maybe I'm missing something as this bench part is not clear to me yet.

pgrange replied:
Sharing some explanations: this is related to the way benchmark works and these numbers help get more insights on the scripts capabilities.

ch1bo committed:
Revert "This does not change test output so keeping it that way"

This reverts commit 65a02a32c1c45ea293b8909712deb8756bce4db2.

ch1bo committed:
Further reduce max supported assets in fanout tests

ch1bo committed:
Fix label in fanout StatSpec

ch1bo committed:
Fix close mutation of snapshot number

It was modifying the input datum instead of the resulting datum and also
using an unrelated slot number to compute the expected contestation
deadline.

pgrange committed:
ADD mutation for the 0 case snapshot in close properties

We might be able to merge close and closeInitial mutations.

Transform the datum instead of recreating everything from scratch

ch1bo committed:
Remove duplicate mutations in CloseInitial

We only keep the mutations which were motivating the separate healthy
case in the first place to reduce duplication.

Ideally, our mutations would be so generic to apply to both
cases (closing with initial and signed snapshots), but we are not there
yet.

pgrange committed:
Mutate only what's needed

This is the closest I can get to mutations that actually mutate a
transaction with what we're intersted in and reuse all the rest
instead of relying on knowledge about how the initial fixture has
been built.

This should help merging some code with Close.hs

pgrange committed:
Merge close and CloseInitial mutations

So that we may converge code, let's regroup both together.

pgrange committed:
Reduce duplication

Extract mutateClosedContestationDeadline to use in both close and
closeInitial mutations

pgrange committed:
Refactor head state mutation

We introduce replaceSomething to ease the mutation of a head state.
This functions are exposed by Mutation.hs as it seems to be the better
place to me right now but maybe we can do better?

Also, we replace some functions which did that and failed when the state
was not an expected one. Now we just return the state unchanged if the
field we want to change does not exist for this state.

It may be a good idea... it also may introduce difficulties in
diagnostic test errors. That is not clear at this point. If ist becomes
a problem, we can sill introduce the errors back easily in the code.

pgrange committed:
FIX test that became useless

The removed test used to check that the validator would fail for a snapshot
with a negative value.

But due to recent changes, it is now OK to have a negative snapshot
number. Although it should fail if we try to use a negative snaphot
number with a non-intiail snapshot.

So we amend the previous test we checked for 0 value with a generated
negative or 0 value.

ch1bo committed:
Generate interesting / mutated deadlines and unify mutation cases

The 0 case can be captured by ensuring the deadline generator covers
this as well.

ch1bo committed:
Refer to spec from close mutation labels

ch1bo committed:
Try to restructure code a bit

ch1bo committed:
Keep a note that fromBuiltinData is increasing script size

pgrange committed:
FIX flaky test

There is a problem with the way we test validity bounds mutations. Most
of the arbitrarily generated values will make the validator fail but
there are no reason to not generate valid transaction with what we had.

To fix that, we distinguish between 3 situations:

1. The lower bound is infinite
2. The upper bound is infinite
3. The upper bound is too high given the contestation deadline

ffakenz commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -74,18 +112,11 @@ healthyOpenHeadTxOut =
-    }
+-- FIXME: This is not a healthy value anyhow related to the 'healthyCloseTx' above
+brokenSlotNo :: SlotNo
+brokenSlotNo = arbitrary `generateWith` 42
```
this is not being used

pgrange replied:
Thank you, that was the point of the previous commits to remove that... oops! forgot to actually remove it ;)

ffakenz commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -139,42 +170,51 @@ healthySignature number = aggregate [sign sk snapshot | sk <- healthySigningKeys
+  | InfiniteUpperBound
+  | -- | See spec: 5.5 rule 4 -> contestationDeadline = upperBound + contestationPeriod
+    MutateContestationDeadline
+  | -- See spec: 5.5. rule 5 -> upperBound - lowerBound <= contestationPeriod
```
idk if it is the best idea to put spec references here

pgrange replied:
Where would you put it? I thought moving these comments to where the mutation code is because I realized I was often scrolling from the code to this comment to check where in the spec to look at.

Or are you suggesting to just get rid of any reference to the spec? If so, what would be your intention there?

ffakenz commented on hydra-node/test/Hydra/Chain/Direct/Contract/Mutation.hs:
```diff
@@ -620,3 +635,71 @@ replacePolicyIdWith originalPolicyId otherPolicyId output =
+  otherState -> otherState
+
+replaceParties :: [Data.Party] -> Head.State -> Head.State
+replaceParties parties = \case
```
all head states have parties as an attribute, so i think would be best to rename the builder functions to reflect the state that is being updated.

pgrange replied:
Well actually the [`Final`](https://github.com/input-output-hk/hydra/blob/gap/move_attributes_from_redeemer_to_datum_for_close_and_contest_tx/hydra-plutus/src/Hydra/Contract/HeadState.hs#L42) state does not have a `parties` field. That is why we have to pattern match against the type of state that we update.

But other than that, this function `replaceParties` would replace the `parties` field of any given state with does have a `parties` field.

Does that help? Otherwise, maybe you can suggest a code change just to help understand the improvement you have in mind.

ffakenz replied:
Oh you are right, we could just extend this function as needed!

Lets go!

ffakenz commented on hydra-node/test/Hydra/Chain/Direct/Contract/Mutation.hs:
```diff
@@ -248,10 +248,10 @@ deriving instance Show SomeMutation
+  | -- | Changes the spent 'Head' script datum to the given value. This modifies
+    -- both the 'DatumHash' in the UTxO context and the map of 'DatumHash' to
+    -- 'Datum' in the transaction's witnesses.
+    ChangeInputHeadDatum Head.State
```
nice renaming!

ffakenz commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -380,32 +400,6 @@ checkFanout utxoHash contestationDeadline numberOfFanoutOutputs ScriptContext{sc
--- | Check that the output datum of this script corresponds to an expected
--- value. Takes care of resolving datum hashes and inline datums.
-checkHeadOutputDatum :: ToData a => ScriptContext -> a -> Bool
-checkHeadOutputDatum ctx d =
```
I still find this check useful to maintain, possibly as the very last validation step for the script.

I know we have made some simplifications that makes this check potentially redundant, but I still see value to keep it at the very end.

wdyt?

ch1bo replied:
We pay for ever bit of code included on-chain, we do not want to keep redundant checks.

pgrange replied:
@ffakenz do you see something that we used to test before that wouldn't be checked by this version of the code? If so, would you share an example? That could be helpful to double-check our mutation tests.

ffakenz replied:
I do not see a counter-example, but this check looks like a catch-all to me, that is the reason why I suggest keeping it.

pgrange replied:
A conversation later in this P.R. mentions that, indeed, we might have a problem with headId and parties being corrupted. I'll resove the conversation here and move it to this [other one](https://github.com/input-output-hk/hydra/pull/677#discussion_r1091213101)

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -368,6 +366,7 @@ closeTx vk closing startSlotNo (endSlotNo, utcTime) openThreadOutput headId =
     CloseWithInitialSnapshot{openUtxoHash} -> openUtxoHash
     CloseWithConfirmedSnapshot{closeUtxoHash} -> closeUtxoHash
 
+  -- FIXME the redeemer should be a maybe as it is useless in the initial close case
```
Feel free to remove this. It seems we will just keep the empty list for this case?

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -833,6 +831,11 @@ observeContestTx utxo tx = do
+  closedSnapshotNumber headDatum =
+    case fromData $ toPlutusData headDatum of
+      Just Head.Closed{snapshotNumber} -> snapshotNumber
+      _ -> error "wrong state in output datum"
```
Should: Not use `error` in pure functions.

Instead this should use the `Maybe` monad to short-cut like other code here.

pgrange replied:
Removed

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -44,27 +44,65 @@ healthyCloseTx =
       , openContestationPeriod = healthyContestationPeriod
       }
+  healthyClosingSnapshot :: ClosingSnapshot
+  healthyClosingSnapshot =
```
No need to call this "healthy" as it cannot be used outside of the `healthyCloseInitialTx` context.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -44,27 +44,65 @@ healthyCloseTx =
+      , openContestationPeriod = healthyContestationPeriod
+      }
+  healthyClosingSnapshot :: ClosingSnapshot
+  healthyClosingSnapshot =
```
No need to call this "healthy" as it cannot be used outside of the `healthyCloseTx` context.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -44,27 +44,65 @@ healthyCloseTx =
+      , signatures = healthySignature healthySnapshotNumber
+      }
+
+healthyCloseInitialTx :: (Tx, UTxO)
```
Would be great if this has a docstring telling how this is different to the other healthy case.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -139,42 +170,51 @@ healthySignature number = aggregate [sign sk snapshot | sk <- healthySigningKeys
+    -- invalid.
+    MutateSnapshotNumberButNotSignature
+  | -- | This test the case when we have a non-initial utxo hash but the snapshot number is 0
+    MutateSnapshotNumberToLessThanZero
```
"LessThanZero" and "snapshot number is 0" do not sound like the same thing?

pgrange replied:
Could you explicit what change in the code you would expect here? That's not clear to me.

ch1bo replied:
Yeah sorry. I expect the documentation to match the label. `is 0` is not the same as `LessThanZero`. It likely diverged and I would expect the documentation to reflect that.

Should: Fix documentation to say `< 0`

pgrange replied:
Of course, my bad.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -187,18 +227,22 @@ genCloseMutation (tx, _utxo) =
+    , SomeMutation (Just "hasBoundedValidity check failed") MutateValidityInterval <$> do
+        (lowerSlotNo, lowerUTCTime) <- genPointInTime
+        (upperSlotNo, upperUTCTime) <- genPointInTime `suchThat` ((> (addContestationPeriod (posixFromUTCTime lowerUTCTime) healthyContestationPeriod)) . posixFromUTCTime . snd)
+        let adjustedContestationDeadline = addContestationPeriod (posixFromUTCTime upperUTCTime) healthyContestationPeriod
```
Instead of using `suchThat` we should be generating a canonical value within the contestation period. For example use an `arbitrary` upper bound, then ~~`choose(0, contestationPeriod)` and subtract that from the upper bound to get the lower.~~ [EDIT: of course we want to have values outside of the bound.. so we just generate a positive integer and add it to the `healthyContestationPeriod` instead.] That way, we never need to `discard` values and not run into endless loops (as we did in the past).

@abailly-iohk didn't you add a coding standard about this when we ran into this last time?

pgrange replied:
I agree that this `suchThat` is problematic but was not able to come with a better solution. Could you write some drafty Haskell code of what you have in mind here?

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -207,50 +251,41 @@ genCloseMutation (tx, _utxo) =
+    pure $ changeHeadOutputDatum (replaceUtxoHash $ toBuiltin mutatedUTxOHash) headTxOut
+
+data CloseInitialMutation
+  = MutateCloseContestationDeadline'
```
Could: add an explanation why this is here, but none of the other mutations? I suspect we want to call this a "regression" test.

pgrange replied:
I actually I think we should test all the other mutations here since we do not  randomly generate either a _normal_ healthy close or an _initial_ healthy close. I'll try to explain that in the code and let as is for this P.R.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Mutation.hs:
```diff
@@ -620,3 +635,71 @@ replacePolicyIdWith originalPolicyId otherPolicyId output =
+      , contestationDeadline
+      , headId
+      }
+  otherState -> otherState
```
I think this is the point where we should think about adding `lens`.

pgrange replied:
I understand that it's an idea for another P.R.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/ContractSpec.hs:
```diff
@@ -114,6 +114,11 @@ spec = parallel $ do
+    prop "is healthy" $
+      propTransactionValidates healthyCloseInitialTx
+    prop "does not survive random adversarial mutations" $
+      propMutation healthyCloseInitialTx genCloseInitialMutation
```
Interestingly, this is currently only one class of generated mutations. No need to change anything here, but it's a great motivating example of using lists of properties instead of `oneof` `Gen SomeMutation`

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -248,46 +257,41 @@ checkClose ::
 
+  (closedSnapshotNumber, closedUtxoHash, closedContestationDeadline) =
+    -- XXX: fromBuiltinData is super big (and also expensive?)
+    case fromBuiltinData @DatumType $ getDatum (continuingDatum ctx) of
```
An interesting CIP which is related: https://github.com/cardano-foundation/CIPs/pull/455

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -248,46 +257,41 @@ checkClose ::
-              , contestationDeadline = makeContestationDeadline cperiod ctx
-              , headId = headPolicyId
-              }
-       in checkHeadOutputDatum ctx expectedOutputDatum
```
This was also checking that the `headId` and `parties` don't change.

Should: do the same as `mustNotChangeParameters` below.

(I added `mustNotChangeParameters` to see whether we would run into bigger scripts / costs if we completely rewrite the checks that was there before.)

pgrange replied:
Added two new mutations to mess around with the headId and the parties to exhibit the issue and fixed. See last commit.

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -350,6 +352,24 @@ checkContest ctx@ScriptContext{scriptContextTxInfo} contestationDeadline parties
+        && headId' == headId
+
+  (contestSnapshotNumber, contestUtxoHash, parties', contestationDeadline', headId') =
+    case fromBuiltinData @DatumType $ getDatum (continuingDatum ctx) of
```
We should probably also mark this `fromBuiltinData` as `XXX` as it is a substantial source of script size.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/StateSpec.hs:
```diff
@@ -389,15 +389,16 @@ forAllFanout action =
             & label ("Fanout size: " <> prettyLength (countAssets $ txOuts' tx))
  where
-  maxSupported = 67
+  maxSupported = 42
```
According to the benchmarks, `52` should be possible (ball-park, not reproducible). Please check again what a maximum supported value for this would be (for `maximumNumberOfParties`)

pgrange replied:
I've set that to 52 and run `cabal bench tx-cost` and observed 50 instead of 46 in _Cost of FanOut Transaction_ array.

Don't know if it matches what you expected.

ch1bo reviewed (commented):
Nearly done.. just some minor things that I pointed out

pgrange committed:
Won't FIX in this P.R.

pgrange committed:
This is not a generally available stuff

No need to call this "healthy" as it cannot be used outside of the healthyCloseInitialTx context.

pgrange committed:
Explain why we would have two healthy close cases to test

pgrange committed:
Explain why we keep this single mutation for initial close

pgrange committed:
FIX one cound change head parameters

One could have tried to change the head parameters during a close,
like change the headId or the parties.

pgrange committed:
Same potential performance issue here

v0d1ch committed:
Remove suchThat and use choose to generate tx validity values

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -36,35 +37,78 @@ import Test.QuickCheck.Instances ()
+
+-- | Healthy close transaction for the specific case were we close a head
+--   with the initial UtxO, that is, no snapshot have been agreed upon and
+--   signed by the head members yet.
```
:tada:

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -183,22 +227,30 @@ genCloseMutation (tx, _utxo) =
+        pure $ ChangeOutput 0 $ changeHeadOutputDatum (replaceParties mutatedParties) headTxOut
+    , SomeMutation (Just "changed parameters") MutateHeadIdInOutput <$> do
+        otherHeadId <- toPlutusCurrencySymbol . headPolicyId <$> arbitrary `suchThat` (/= Fixture.testSeedInput)
+        pure $ ChangeOutput 0 $ changeHeadOutputDatum (replaceHeadId otherHeadId) headTxOut
```
Nice work

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -183,22 +227,30 @@ genCloseMutation (tx, _utxo) =
+        mutatedParties <- arbitrary `suchThat` (/= healthyOnChainParties)
+        pure $ ChangeOutput 0 $ changeHeadOutputDatum (replaceParties mutatedParties) headTxOut
+    , SomeMutation (Just "changed parameters") MutateHeadIdInOutput <$> do
+        otherHeadId <- toPlutusCurrencySymbol . headPolicyId <$> arbitrary `suchThat` (/= Fixture.testSeedInput)
```
Could: exclude the healthy policy id used above more explicitly by using `` `suchThat` (/= Fixture.testPolicyId)``

pgrange replied:
I agree, this it not clear that testSeedInput is actually the policyId.

The thing is, there is no instance of arbitrary for a `PolicyId` so I wouldn't know how to apply suchThat to a PolicyId so I'll keep it as is even though it's not perfect, I agree.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -207,50 +259,55 @@ genCloseMutation (tx, _utxo) =
+-- least we keep this regression test as we stumbled upon problems with the following case.
+-- The nice thing to do would probably to generate either "normal" healthy close or
+-- or initial healthy close and apply all the mutation to it but we did'nt manage to do that
+-- rightaway.
```
:heart:

ch1bo reviewed (approved):
Only a very minor thing left and the tests seem to not agree on the `maxSupported` assets as it fails with `45` utxos on the fanout.

Feel free to merge after fixing this so I approve.

pgrange committed:
Back to working value
